---
title: "introduction"
output: html_document
vignette: >
  %\VignetteIndexEntry{introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# load package
```{r setup}
#devtools::install_github("Czh3/iSpatial", force = T)
library(iSpatial)
```

# infer merFISH data:
```{r example}
# scRNA-seq data
data(NA_scRNA)

# merFISH data
data(NA_merFISH)

library(Seurat)
library(ggplot2)

# infer spatial transcriptome
NA_merFISH_iSpatial = iSpatial::iSpatial(NA_merFISH, NA_scRNA)


# Gene "Acbd7" is not in the merFISH targeted genes
"Acbd7" %in% rownames(NA_merFISH)

# After inferring, we can get the spatial expression of Acbd7
SpatialFeaturePlot(NA_merFISH_iSpatial, features = "Acbd7", stroke=0.1, slot="data") +
  theme(legend.position = "right") +
  scale_fill_gradientn(colours = c(rainbow(10)[5:10], "red"))

# Find spatial variable genes
SVGs_infer = spatial_variable_genes(NA_merFISH_iSpatial, "enhanced", n.core = 10)
SVGs_infer_sig = SVGs_infer[SVGs_infer$p.value < 0.001, ]

g = ggplot2::ggplot(SVGs_infer, aes(pct.expr, -log10(p.value), gene = gene, color = -log10(p.value))) +
  geom_point()+
  theme_classic()
plotly::ggplotly(g)

# cluster genes based on spatial expression pattern
gene_cluster = suppressWarnings(cluster_spatial_expr_pattern(NA_merFISH_iSpatial, "enhanced", 
                             feature = SVGs_infer_sig$gene, n.cluster = 12,
                             ncol.plot = 3, bin = 20, n.core = 10))

# genes in cluster 7
names(gene_cluster[gene_cluster==7])

# plot
spatial_signature_plot(NA_merFISH_iSpatial, 
                       gene.set = names(gene_cluster[gene_cluster==7]),
                       set.name = "Class 7")

```

```{r}
sessionInfo()
```

